File: LeetCodePractice/src/DP/BestTimeToBuyAndSellStockIV.java
package DP;
/*Say you have an array for which the ith element is the price of a given stock on day i.

Design an algorithm to find the maximum profit. You may complete at most k transactions.

Example
Given prices = [4,4,6,1,1,4,2,5], and k = 2, return 6.

Challenge
O(nk) time.

Notice
You may not engage in multiple transactions at the same time (i.e., you must sell the stock before you buy again).
*/
public class BestTimeToBuyAndSellStockIV {
	/**
     * @param K: An integer
     * @param prices: An integer array
     * @return: Maximum profit
     */
    public int maxProfit(int K, int[] prices) {
        // write your code here
    	if(prices == null || prices.length == 0){
            return 0;
        }
        int len = prices.length;
        if(K >= len / 2){
        	return sellStockII(prices);
        }
        int[][] dp = new int[K + 1][len];
        //DP: mem(i,j) is the max profit for up to i transactions by time j (0<=i<=K, 0<=j< len).
        for(int i = 1; i <= K; i++){
        	int balance = -prices[0];
        	for(int j = 1; j < len; j++){
        		dp[i][j] = Math.max(dp[i][j - 1], balance + prices[j]);
        		balance = Math.max(balance, dp[i - 1][j - 1] - prices[j]);
        		//accumulate means the maximum profit of just doing at most i-1 transactions, using at most first j-1 prices, and buying the stock at price[j]
        	}
        }
        return dp[K][len - 1];
    }
    private int sellStockII(int[] prices){
    	int res = 0;
    	int len = prices.length;
    	for(int i = 1; i < len; i++){
    	    if(prices[i] > prices[i - 1]){
    	        res += prices[i] - prices[i - 1];
    	    }
    	}
    	return res;
    }
}

File: LeetCodePython/BestTimeToBuyAndSellStockIV.py
"""
You are given an integer array prices where prices[i] is the price of a given stock on the ith day, and an integer k.

Find the maximum profit you can achieve. You may complete at most k transactions.

Note: You may not engage in multiple transactions simultaneously (i.e., you must sell the stock before you buy again).

Example 1:
Input: k = 2, prices = [2,4,1]
Output: 2
Explanation: Buy on day 1 (price = 2) and sell on day 2 (price = 4), profit = 4-2 = 2.

Example 2:
Input: k = 2, prices = [3,2,6,5,0,3]
Output: 7
Explanation: Buy on day 2 (price = 2) and sell on day 3 (price = 6), profit = 6-2 = 4. Then buy on day 5 (price = 0) and sell on day 6 (price = 3), profit = 3-0 = 3.

Constraints:
0 <= k <= 100
0 <= prices.length <= 1000
0 <= prices[i] <= 1000

Challenge
O(nk) time.

analysis:
dp: mem(i,j) is the max profit for up to i transactions by time j (0<=i<=K, 0<=j< len).
"""
from typing import List


class BestTimeToBuyAndSellStockIV:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        if not prices:
            return 0

        def sell_stock_II():
            res = 0
            for i in range(1, len(prices)):
                if prices[i - 1] < prices[i]:
                    res += prices[i] - prices[i - 1]
            return res

        if k >= len(prices) // 2:
            return sell_stock_II()

        dp = [[0] * len(prices) for _ in range(k + 1)]
        for i in range(1, k + 1):
            hold = -prices[0]
            for j in range(1, len(prices)):
                dp[i][j] = max(dp[i][j - 1], hold + prices[j])
                # accumulate means the maximum profit of just doing at most i-1 transactions,
                # using at most first j-1 prices, and buying the stock at price[j]
                hold = max(hold, dp[i - 1][j - 1] - prices[j])
        return dp[k][-1]